src/subdataframe/subdataframe.jl

```diff
@@ -1,154 +1,208 @@
+@inline function SubDataFrame(parent::DataFrame, rows::AbstractVector{Bool}, cols)
     if length(rows) != nrow(parent)
         throw(ArgumentError("invalid length of `AbstractVector{Bool}` row index" *
                             " (got $(length(rows)), expected $(nrow(parent)))"))
     end
-    return SubDataFrame(parent, findall(rows))
+    return SubDataFrame(parent, findall(rows), cols)
 end
 
-function SubDataFrame(parent::DataFrame, rows::AbstractVector)
+@inline function SubDataFrame(parent::DataFrame, rows::AbstractVector, cols)
     if !all(x -> (x isa Integer) && !(x isa Bool), rows)
         throw(ArgumentError("only `Integer` indices are accepted in `rows`"))
     end
-    return SubDataFrame(parent, convert(Vector{Int}, rows))
+    return SubDataFrame(parent, convert(Vector{Int}, rows), cols)
 end
 
-function SubDataFrame(sdf::SubDataFrame, rowinds)
-    return SubDataFrame(parent(sdf), rows(sdf)[rowinds])
+getparentcols(sdf::SubDataFrame, idx::Union{Integer, AbstractVector{<:Integer}}) =
+    getfield(sdf, :cols)[idx]
+
+function getparentcols(sdf::SubDataFrame, idx::Symbol)
+    parentcols = index(parent(sdf))[idx]
+    getfield(sdf, :remap)[parentcols] == 0 && throw(KeyError("$idx not found"))
+    return parentcols
 end
 
-SubDataFrame(sdf::SubDataFrame, rowinds::Colon) = sdf
+getparentcols(sdf::SubDataFrame, idx::AbstractVector{Symbol}) =
+    [getparentcols(sdf, i) for i in idx]
+
+getparentcols(sdf::SubDataFrame, ::Colon) = getfield(sdf, :cols)
+
+SubDataFrame(sdf::SubDataFrame, rowind, cols) =
+    SubDataFrame(parent(sdf), rows(sdf)[rowind], getparentcols(sdf, cols))
+SubDataFrame(sdf::SubDataFrame, ::Colon, cols) =
+    SubDataFrame(parent(sdf), rows(sdf), getparentcols(sdf, cols))
+SubDataFrame(sdf::SubDataFrame, ::Colon, ::Colon) = sdf
 
-Base.parent(sdf::SubDataFrame) = getfield(sdf, :parent)
-Base.parentindices(sdf::SubDataFrame) = (rows(sdf), axes(parent(sdf), 2))
 rows(sdf::SubDataFrame) = getfield(sdf, :rows)
+Base.parent(sdf::SubDataFrame) = getfield(sdf, :parent)
+Base.parentindices(sdf::SubDataFrame) = (rows(sdf), getfield(sdf, :cols))
 
 Base.view(adf::AbstractDataFrame, colinds) = view(adf, :, colinds)
 Base.view(adf::AbstractDataFrame, rowinds, colind::ColumnIndex) =
     view(adf[colind], rowinds)
 Base.view(adf::AbstractDataFrame, rowinds, colind::Bool) =
     throw(ArgumentError("invalid column index $colind of type `Bool`"))
 Base.view(adf::AbstractDataFrame, rowinds, colinds) =
-    SubDataFrame(adf[colinds], rowinds)
-Base.view(adf::AbstractDataFrame, rowinds, ::Colon) = SubDataFrame(adf, rowinds)
+    SubDataFrame(adf, rowinds, colinds)
 
 ##############################################################################
 ##
 ## AbstractDataFrame interface
 ##
 ##############################################################################
 
-index(sdf::SubDataFrame) = index(parent(sdf))
+struct SubIndex{T,S} <: AbstractIndex
+    sdf::SubDataFrame{T,S}
```

Comment:
Why not have a single parameter for the `SubDataFrame`?

Or probably more logical: store the parent `Index` and the `cols`. Then you can define this in index.jl. That's more logical since `DataFrame` is based on `Index`, not the other way around, so `SubIndex` shouldn't be based on `SubDataFrame`.

Reply:
OK - moved